api/app/Http/Middleware/AuthMiddleware.php

```diff
@@ -0,0 +1,58 @@
+<?php namespace App\Http\Middleware;
+
+use Closure;
+use Tymon\JWTAuth\JWTAuth;
+use Illuminate\Http\Request;
+use App\Exceptions\ForbiddenException;
+
+class AuthMiddleware
+{
+    /**
+     * JWT Auth
+     *
+     * @var JWTAuth
+     */
+    protected $jwtAuth;
+
+    /**
+     * Assign dependencies.
+     *
+     * @param  JWTAuth  $jwtAuth
+     * @return void
+     */
+    public function __construct(JWTAuth $jwtAuth)
+    {
+        $this->jwtAuth = $jwtAuth;
+    }
+
+    /**
+     * Handle an incoming request.
+     *
+     * @param  Request  $request
+     * @param  Closure  $next
+     * @return mixed
+     */
+    public function handle(Request $request, Closure $next)
+    {
+        // Get possible x amount of parameters passed to the middleware
+        $userTypes = array_slice(func_get_args(), 2);
+
+        $token = $this->jwtAuth->getTokenFromRequest($request);
+        $user = $this->jwtAuth->getUser($token);
+
+        // Check the restriction types
+        if (in_array('admin', $userTypes) and $user->user_type == 'admin') {
```

Comment:
@xiphiaz Quick initial response regarding this. I've used the entrust package in the past, and I'd personally not select it again. Many things I didn't like about it, and quite db heavy. Last time I used it we kept working at phasing it out and replacing it with a service based permission system not having to hit different db tables unnecessary.

The lock laravel looks interesting though, so I'm going to take a closer lock at that one and see how it works and what it might offer.

Looking forward. Except for permission based on model property. What else will be required of the permission system?